airflow/models/taskinstance.py

```diff
@@ -186,6 +186,193 @@ class TaskReturnCode(Enum):
+    ti.refresh_from_db(session=session)
+
+    ti.job_id = job_id
+    ti.hostname = get_hostname()
+    ti.pid = os.getpid()
+    if not test_mode:
+        TaskInstance.save_to_db(ti=ti, session=session)
+    actual_start_date = timezone.utcnow()
+    Stats.incr(f"ti.start.{ti.task.dag_id}.{ti.task.task_id}", tags=ti.stats_tags)
+    # Same metric with tagging
+    Stats.incr("ti.start", tags=ti.stats_tags)
+    # Initialize final state counters at zero
+    for state in State.task_states:
+        Stats.incr(
+            f"ti.finish.{ti.task.dag_id}.{ti.task.task_id}.{state}",
+            count=0,
+            tags=ti.stats_tags,
+        )
+        # Same metric with tagging
+        Stats.incr(
+            "ti.finish",
+            count=0,
+            tags={**ti.stats_tags, "state": str(state)},
+        )
+    with set_current_task_instance_session(session=session):
+        ti.task = ti.task.prepare_for_execution()
+        context = ti.get_template_context(ignore_param_exceptions=False, session=session)
+
+        try:
+            if not mark_success:
+                TaskInstance._execute_task_with_callbacks(
+                    self=ti,  # type: ignore[arg-type]
+                    context=context,
+                    test_mode=test_mode,
+                    session=session,
+                )
+            if not test_mode:
+                ti.refresh_from_db(lock_for_update=True, session=session)
+            ti.state = TaskInstanceState.SUCCESS
+        except TaskDeferred as defer:
+            # The task has signalled it wants to defer execution based on
+            # a trigger.
+            if raise_on_defer:
+                raise
+            ti.defer_task(exception=defer, session=session)
+            ti.log.info(
+                "Pausing task as DEFERRED. dag_id=%s, task_id=%s, execution_date=%s, start_date=%s",
+                ti.dag_id,
+                ti.task_id,
+                _date_or_empty(task_instance=ti, attr="execution_date"),
+                _date_or_empty(task_instance=ti, attr="start_date"),
+            )
+            if not test_mode:
+                _add_log(event=ti.state, task_instance=ti, session=session)
+                TaskInstance.save_to_db(ti=ti, session=session)
+            return TaskReturnCode.DEFERRED
+        except AirflowSkipException as e:
+            # Recording SKIP
+            # log only if exception has any arguments to prevent log flooding
+            if e.args:
+                ti.log.info(e)
+            if not test_mode:
+                ti.refresh_from_db(lock_for_update=True, session=session)
+            ti.state = TaskInstanceState.SKIPPED
+            _run_finished_callback(callbacks=ti.task.on_skipped_callback, context=context)
+            TaskInstance.save_to_db(ti=ti, session=session)
+        except AirflowRescheduleException as reschedule_exception:
+            ti._handle_reschedule(actual_start_date, reschedule_exception, test_mode, session=session)
```

Comment:
I think `_handle_reschedule` should be simply turned into an `@internal_api` call same as `handle_failure` (and then `save_to_db` would not be needed at all as well - because merge/commit is already done inside `_handle_reschedule`. That would also solve the `todo:` below where this was supposed to be done.

Reply:
yes it needs to be done.

Reply:
done